tests/unit/test_actions.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Jenkins-k8s charm actions unit tests."""
+
+import secrets
+
+# Need access to protected functions for testing
+import typing
+from unittest.mock import MagicMock
+
+import ops
+import pytest
+
+import charm
+import jenkins
+from charm import JenkinsK8sOperatorCharm
+
+from .types_ import HarnessWithContainer
+
+
+def test_on_get_admin_password_action_container_not_ready(
+    harness_container: HarnessWithContainer,
+):
+    """
+    arrange: given a jenkins container that is not connectable.
+    act: when get-admin-password action is run.
+    assert: the event is failed.
+    """
+    harness_container.harness.set_can_connect(
+        harness_container.harness.model.unit.containers["jenkins"], False
+    )
+    mock_event = MagicMock(spec=ops.ActionEvent)
+    harness_container.harness.begin()
+
+    jenkins_charm = typing.cast(JenkinsK8sOperatorCharm, harness_container.harness.charm)
+    jenkins_charm.actions_observer.on_get_admin_password(mock_event)
+
+    mock_event.fail.assert_called_once()
+
+
+def test_on_get_admin_password_action(
+    harness_container: HarnessWithContainer, admin_credentials: jenkins.Credentials
+):
+    """
+    arrange: given a jenkins container.
+    act: when get-admin-password action is run.
+    assert: the correct admin password is returned.
+    """
+    mock_event = MagicMock(spec=ops.ActionEvent)
+    harness_container.harness.begin()
+
+    jenkins_charm = typing.cast(JenkinsK8sOperatorCharm, harness_container.harness.charm)
+    jenkins_charm.actions_observer.on_get_admin_password(mock_event)
+
+    mock_event.set_results.assert_called_once_with({"password": admin_credentials.password})
+
+
+def test_on_rotate_credentials_action_container_not_ready(
```

Comment:
@benhoyt also something to consider, it seems that not being able to connect to the container would probably fail many of the action events, something to consider whether the framework could provide guarantees, a decorator or something that makes this less intensive to code and then to test within charms for each of the actions

Reply:
It seems to me the test is doing too much here. There's no need to call `set_can_connect` with `False` as that's the default for containers under `Harness` anyway (when using `begin` and not `begin_with_initial_hooks`).

Additionally, I just looked at the code for this action:

```python
    def _on_get_admin_password(self, event: ops.ActionEvent) -> None:
        container = self.unit.get_container(self.state.jenkins_service_name)
        if not container.can_connect():
            event.defer()
            return
        credentials = jenkins.get_admin_credentials(container)
        event.set_results({"password": credentials.password_or_token})
```

It doesn't seem right at all to use `defer()` in an `ActionEvent`. In fact, `ActionEvent.defer` always raises a `RuntimeError` -- so I'm not sure how this code path is working at all (maybe it's not and the test isn't actually testing what we think it is). We definitely shouldn't use `defer` inside an action; this should be changed to an `event.fail()` or similar (or just let it error?).

Reply:
The test has been simplified with harness (`harness_with_container` fixture by default enables jenkins container to be ready to reduce repeating set_can_connect).

For the later code snippet, the event defer has been removed a few PRs back :)
Thank you for the review!